skimage/future/_thin_plate_splines.py

```diff
@@ -0,0 +1,323 @@
+        )
+        y_warp = self._spline_function(
+            coords[:, 0], coords[:, 1], radial_dist, coeffs[:, 1]
+        )
+        return np.vstack([x_warp, y_warp]).T
+
+    @property
+    def inverse(self):
+        raise NotImplementedError("This is yet to be implemented.")
+
+    def estimate(self, src, dst):
+        """Estimate optimal spline mappings between source and destination points.
+
+        Parameters
+        ----------
+        src : (N, 2) array_like
+            Control points at source coordinates.
+        dst : (N, 2) array_like
+            Control points at destination coordinates.
+
+        Returns
+        -------
+        success: bool
+            True indicates that the estimation was successful.
+
+        Notes
+        -----
+        The number N of source and destination points must match.
+        """
+
+        check_nD(src, 2)
+        check_nD(dst, 2)
+
+        if len(src) < 3 or len(dst) < 3:
+            raise ValueError(
+                "There should be at least 3 points in both sets (source and destination)."
+            )
+
+        if src.shape != dst.shape:
+            raise ValueError(
+                "The shape of source and destination coordinates must match."
+            )
+
+        self.src = src
+        n, d = src.shape
+
+        dist = sp.spatial.distance.cdist(self.src, self.src)
+        K = _radial_basis_kernel(dist)
+        P = np.hstack([np.ones((n, 1)), src])
+        L = np.zeros((n + 3, n + 3), dtype=np.float32)
+        L[:n, :n] = K
+        L[:n, -3:] = P
+        L[-3:, :n] = P.T
+        V = np.concatenate([dst, np.zeros((d + 1, d))])
+        self.spline_mappings = np.dot(np.linalg.inv(L), V)
+        return True
+
+    def _radial_distance(self, x, y):
+        """Compute the radial distance."""
+        Pi_x = self.src[:, 0]
+        Pi_y = self.src[:, 1]
+
+        dx = Pi_x[:, np.newaxis] - x
+        dy = Pi_y[:, np.newaxis] - y
+        r = np.sqrt(dx**2 + dy**2)
+        radial_dist = _radial_basis_kernel(r)
+        return radial_dist
+
+    def _spline_function(self, x, y, radial_dist, coeffs):
+        """Estimate the spline function in X and Y directions."""
+        w = coeffs[:-3]
+        a1, ax, ay = coeffs[-3:]
+        summation = np.dot(w, radial_dist)
+        return a1 + ax * x + ay * y + summation
+
+
+def _radial_basis_kernel(r):
+    """Compute basis kernel for thin-plate splines.
+
+    Parameters
+    ----------
+    r : (4, N) ndarray
+        Input array representing the norm distance between interlandmark
+        distances for the source (control) points and based on the (x,y) coordinates
+        for each of these points.
+
+    Returns
+    -------
+    U : (4, N) ndarray
+        Calculated kernel function U.
+    """
+    _small = 1e-8  # Small value to avoid divide-by-zero
+    r_sq = r**2
+    U = np.where(r == 0.0, 0.0, r_sq * np.log(r_sq + _small))
+    return U
+
+
+def tps_warp(
```

Comment:
See https://github.com/scikit-image/scikit-image/pull/7040#discussion_r1592084562 above on the parameter `grid_scaling`.

Reply:
From the same discussion: let's go with `warp` instead of adding `tpw_warp`. We can look into whether we can integrate the `grid_scaling` optimization in `warp` itself but it is not that essential to memory optimization (@stefanv can explain that better than me). 

I'm not sure when I will get time to look into implementing that right now. @decorouz, feel welcome give it a try if you have time and incentive. :D

Reply:
The [short mathematical answer](https://math.stackexchange.com/a/1978293/656185).
The [longer mathematical unpacking](https://gregorygundersen.com/blog/2020/12/09/matrix-inversion/).

Reply:
For each pixel in the output image, we need to find out where in the input image to "fetch" it from. We can do that by computing a coordinate transformation for every pixel. That can be done per pixel (very low memory consumption), but in Python we typically vectorize it, i.e. use a bunch of memory to store the coordinate map (output coord -> input coord).

I haven't looked at this carefully, but I suspect the optimization used here is to generate a smaller coordinate map, and to then trick `map_coordinate` into interpolating the coordinates as needed. Less accurated, but smaller memory consumption, and perhaps, due to caching, a little bit faster.